myia/prim/shape_inferrers.py

```diff
@@ -0,0 +1,157 @@
+
+
+class DefaultShapeInferrer(Inferrer):
+    """Shape inferrer for all primitves that don't take arrays."""
+
+    def __init__(self, engine):
+        """Initialize the DefaultShapeInferrer."""
+        super().__init__(engine, 'default_shape_inferrer')
+
+    async def __call__(self, *args):
+        """Since no arrays are involved, the shape is always ()."""
+        return ()
+
+    def provably_equivalent(self, other):
+        """This is always equal to itself."""
+        return type(self) == type(other)
+
+
+class ShapeTrack(Track):
+    """Infer the shape of a constant."""
+
+    def __init__(self, engine, name, *,
+                 constructors=shape_inferrer_constructors):
+        """Intialize a ShapeTrack."""
+        super().__init__(engine, name)
+        self.constructors = constructors
+
+    def from_value(self, v, context):
+        """Infer the shape of a constant."""
+        if isinstance(v, Primitive):
+            if v in self.constructors:
+                return self.constructors[v](self.engine)
+            else:
+                return DefaultShapeInferrer(self.engine)
+        elif isinstance(v, Graph):
+            return GraphInferrer(self.engine, 'shape', v, context)
+        else:
+            return getattr(v, 'shape', ())
+
+
+shape_inferrer = partial(register_inferrer,
+                         constructors=shape_inferrer_constructors)
+
+
+@shape_inferrer(P.return_, nargs=1)
+async def infer_shape_return(engine, v):
+    """Infer the shape of return."""
+    return await engine.get('shape', v)
+
+
+@shape_inferrer(P.if_, nargs=3)
+async def infer_shape_if(engine, cond, tb, fb):
+    """Infer the shape of if."""
+    tb_inf = await engine.get('shape', tb)
+    fb_inf = await engine.get('shape', fb)
+    v = await engine.get('value', cond)
+    if v is True:
+        # We only visit the first branch if the condition is provably true
+        return await tb_inf()
+    elif v is False:
+        # We only visit the second branch if the condition is provably false
+        return await fb_inf()
+    elif v is ANYTHING:
+        # The first branch to finish will return immediately. When the other
+        # branch finishes, its result will be checked against the other.
+        return await engine.assert_same('shape', tb_inf(), fb_inf())
+
+
+@shape_inferrer(P.map_array, nargs=2)
+async def infer_shape_map_array(engine, fn, ary):
+    """Infer the shape of map_array."""
+    return await engine.get('shape', ary)
+
+
+@shape_inferrer(P.scan_array, nargs=4)
+async def infer_shape_scan_array(engine, fn, init, ary, ax):
+    """Infer the shape of scan_array."""
+    return await engine.get('shape', ary)
+
+
+@shape_inferrer(P.reduce_array, nargs=4)
+async def infer_shape_reduce_array(engine, fn, init, ary, ax):
+    """Infer the shape of reduce_array."""
+    shp = await engine.get('shape', ary)
+    ax_v = await engine.get('value', ax)
+    if ax_v == ANYTHING:
+        return (None,) * (len(shp) - 1)
```

Comment:
Could we use ANYTHING or some special object instead of None?

Reply:
Probably